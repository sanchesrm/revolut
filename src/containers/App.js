import React, { Component } from "react";
import { Container, Header, Image, Button, Message } from "semantic-ui-react";
import CurrencyContent from "../components/CurrencyContent";
import PropTypes from "prop-types";
import { fetchRates, getRatesExchanges } from "../actions/Rates";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import Logo from "../assets/imgs/logo.png";

const styles = {
  buttonStyle: {
    width: "300px",
    height: "50px",
    marginTop: "25px",
    backgroundImage:
      "linear-gradient(to right, rgba(0, 128, 0, 0.5), rgba(0, 128, 0, 0.7)",
    fontSize: "20px"
  },
  divStyle: {
    position: "absolute",
    width: "100%",
    height: "100%",
    backgroundColor: "#1881ec"
  }
};

export class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      choosenCurrency: ["GBP", "USD"],
      currencyAmounts: {
        GBP: this.randomCurrencyAmounts(),
        EUR: this.randomCurrencyAmounts(),
        USD: this.randomCurrencyAmounts()
      },
      inputValues: { 0: "0.00", 1: "0.00" },
      hideError: true,
      errorMessage: ""
    };
  }

  componentDidMount() {
    this.props.fetchRates(this.state.choosenCurrency);
    setInterval(() => {
      this.props.fetchRates(this.state.choosenCurrency);
    }, 10000);
  }

  randomCurrencyAmounts = () => {
    return parseFloat((Math.random() * (100 - 0) + 0).toFixed(2));
  };

  handleCurrencyChoose = (name, position) => {
    let choosenCurrency = this.state.choosenCurrency;
    if (!choosenCurrency.includes(name)) {
      choosenCurrency[position] = name;
      this.setState({ choosenCurrency }, () => {
        if (this.props.RatesReducer.exchangesRates) {
          this.props.getRatesExchanges(
            this.state.choosenCurrency,
            this.props.RatesReducer
          );
        } else {
          this.showErrorMessage(
            "It was not possible to get the currencies rates. Check your internet connection"
          );
        }
      });
    }
  };

  changeHandler = (value, position) => {
    if (this.props.RatesReducer.exchangesRates) {
      const oppositePosition = !position ? 1 : 0;
      let { inputValues } = this.state;
      inputValues[position] = parseFloat(value);
      inputValues[oppositePosition] =
        -1 * value * this.props.RatesReducer.exchangesRates[position];

      this.setState({ inputValues });
    } else {
      this.showErrorMessage(
        "It was not possible to get the currencies rates. Check your internet connection"
      );
    }
  };

  exchangeClickHandler = () => {
    let { choosenCurrency, currencyAmounts, inputValues } = this.state;

    if (!this.props.RatesReducer.exchangesRates) {
      this.showErrorMessage("Type a value to exchange");
    } else if (inputValues[0] === "0.00" || inputValues[0] === "0.00") {
      this.showErrorMessage(
        "It was not possible to get the currencies rates. Check your internet connection"
      );
    } else {
      let firstValue = currencyAmounts[choosenCurrency[0]] + inputValues[0];
      let secondValue = currencyAmounts[choosenCurrency[1]] + inputValues[1];

      firstValue = firstValue.toFixed(2);
      secondValue = secondValue.toFixed(2);

      if (firstValue < 0 || secondValue < 0) {
        this.showErrorMessage("Operation not permitted");
      } else {
        currencyAmounts[choosenCurrency[0]] = parseFloat(firstValue);
        currencyAmounts[choosenCurrency[1]] = parseFloat(secondValue);

        inputValues = { 0: "0.00", 1: "0.00" };

        this.setState({
          currencyAmounts,
          inputValues
        });
      }
    }
  };

  showErrorMessage = errorMessage => {
    this.setState(
      {
        hideError: false,
        errorMessage
      },
      () => {
        setTimeout(() => {
          this.setState({
            hideError: true,
            errorMessage: ""
          });
        }, 3000);
      }
    );
  };

  render() {
    const {
      inputValues,
      choosenCurrency,
      currencyAmounts,
      hideError,
      errorMessage
    } = this.state;
    return (
      <div style={styles.divStyle}>
        <Container
          textAlign="center"
          style={{ marginTop: "25px", width: "600px" }}
        >
          <Header textAlign="center" style={{ marginBottom: "15px" }}>
            <Image src={Logo} style={{ width: "150px" }} />
          </Header>

          <CurrencyContent
            position={0}
            handleCurrencyChoose={this.handleCurrencyChoose}
            choosenCurrency={choosenCurrency}
            amount={currencyAmounts[choosenCurrency[0]]}
            changeHandler={this.changeHandler}
            inputValue={inputValues[0]}
          />
          <CurrencyContent
            position={1}
            handleCurrencyChoose={this.handleCurrencyChoose}
            choosenCurrency={choosenCurrency}
            amount={currencyAmounts[choosenCurrency[1]]}
            changeHandler={this.changeHandler}
            inputValue={inputValues[1]}
          />
          <Button
            style={styles.buttonStyle}
            content="Exchange"
            onClick={this.exchangeClickHandler}
            primary
          />
          <Message
            style={{ marginTop: "20px" }}
            color="red"
            error={true}
            hidden={hideError}
          >
            {errorMessage}
          </Message>
        </Container>
      </div>
    );
  }
}

App.propTypes = {
  fetchRates: PropTypes.func.isRequired,
  getRatesExchanges: PropTypes.func.isRequired
};

export const mapStateToProps = ({ RatesReducer }) => {
  return { RatesReducer };
};

export const mapDispatchToProps = dispatch => {
  return bindActionCreators({ fetchRates, getRatesExchanges }, dispatch);
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);
