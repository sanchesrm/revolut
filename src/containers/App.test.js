import Adapter from "enzyme-adapter-react-16";
import React from "react";
import { configure, shallow } from "enzyme";
import { App, mapStateToProps } from "./App";

describe("Login Component", () => {
  let props;
  beforeEach(() => {
    configure({ adapter: new Adapter() });
    props = {
      fetchRates: jest.fn(() => Promise.resolve()),
      getRatesExchanges: jest.fn(() => Promise.resolve()),
      RatesReducer: {}
    };
  });

  test("should generate random number", () => {
    expect.assertions(1);

    const wrapper = shallow(<App {...props} />);
    const randomNumber = wrapper.instance().randomCurrencyAmounts();

    expect(typeof randomNumber === "number").toEqual(true);
  });

  test("should handle currency change", () => {
    expect.assertions(3);

    props = { ...props, RatesReducer: { exchangesRates: [] } };

    const wrapper = shallow(<App {...props} />);
    const getRatesExchangesMethod = jest.spyOn(props, "getRatesExchanges");

    expect(wrapper.instance().state.choosenCurrency).toEqual(["GBP", "USD"]);

    wrapper.instance().handleCurrencyChoose("EUR", 1);

    expect(wrapper.instance().state.choosenCurrency).toEqual(["GBP", "EUR"]);
    expect(getRatesExchangesMethod).toHaveBeenCalled();
  });

  test("should handle currency change but it's already on the array", () => {
    expect.assertions(3);

    const wrapper = shallow(<App {...props} />);
    const getRatesExchangesMethod = jest.spyOn(props, "getRatesExchanges");

    expect(wrapper.instance().state.choosenCurrency).toEqual(["GBP", "USD"]);

    wrapper.instance().handleCurrencyChoose("GBP", 1);

    expect(wrapper.instance().state.choosenCurrency).toEqual(["GBP", "USD"]);
    expect(getRatesExchangesMethod).not.toHaveBeenCalled();
  });

  test("should handle currency change but it shouldn't have a exchangesRates array", () => {
    expect.assertions(3);

    props = { ...props, RatesReducer: {} };

    const wrapper = shallow(<App {...props} />);
    const showErrorMessageMethod = jest.spyOn(
      wrapper.instance(),
      "showErrorMessage"
    );

    expect(wrapper.instance().state.choosenCurrency).toEqual(["GBP", "USD"]);

    wrapper.instance().handleCurrencyChoose("EUR", 1);

    expect(wrapper.instance().state.choosenCurrency).toEqual(["GBP", "EUR"]);
    expect(showErrorMessageMethod).toHaveBeenCalled();
  });

  test("should handle value typing", () => {
    expect.assertions(2);
    props = {
      ...props,
      RatesReducer: {
        exchangesRates: [1, 2]
      }
    };

    const wrapper = shallow(<App {...props} />);

    expect(wrapper.instance().state.inputValues).toEqual({
      0: "0.00",
      1: "0.00"
    });

    wrapper.instance().changeHandler(5, 1);

    expect(wrapper.instance().state.inputValues).toEqual({ "0": -10, "1": 5 });
  });

  test("should handle value typing but it doesn't have exchangesRates array", () => {
    expect.assertions(2);
    props = {
      ...props,
      RatesReducer: {}
    };

    const wrapper = shallow(<App {...props} />);
    const showErrorMessageMethod = jest.spyOn(
      wrapper.instance(),
      "showErrorMessage"
    );

    expect(wrapper.instance().state.inputValues).toEqual({
      0: "0.00",
      1: "0.00"
    });

    wrapper.instance().changeHandler(5, 1);
    expect(showErrorMessageMethod).toHaveBeenCalled();
  });

  test("should handle value typing - position 0", () => {
    expect.assertions(2);
    props = {
      ...props,
      RatesReducer: {
        exchangesRates: [1, 2]
      }
    };

    const wrapper = shallow(<App {...props} />);

    expect(wrapper.instance().state.inputValues).toEqual({
      0: "0.00",
      1: "0.00"
    });

    wrapper.instance().changeHandler(5, 0);

    expect(wrapper.instance().state.inputValues).toEqual({ "0": 5, "1": -5 });
  });

  test("should handle click and values === 0.00", () => {
    expect.assertions(2);

    const wrapper = shallow(<App {...props} />);
    wrapper.instance().setState({
      currencyAmounts: {
        GBP: 10,
        USD: 20
      },
      inputValues: { 0: 30.0, 1: -30.0 }
    });
    const showErrorMessageMethod = jest.spyOn(
      wrapper.instance(),
      "showErrorMessage"
    );

    wrapper.instance().exchangeClickHandler();

    expect(wrapper.instance().state.hideError).toEqual(false);
    expect(showErrorMessageMethod).toHaveBeenCalledWith(
      "Type a value to exchange"
    );
  });

  test("should handle click and value < 0", () => {
    expect.assertions(2);

    const wrapper = shallow(<App {...props} />);
    wrapper.instance().setState({
      currencyAmounts: {
        GBP: 10,
        USD: 20
      },
      inputValues: { 0: 40.0, 1: -40.0 }
    });
    wrapper.instance().forceUpdate();
    const showErrorMessageMethod = jest.spyOn(
      wrapper.instance(),
      "showErrorMessage"
    );

    wrapper.instance().exchangeClickHandler();

    expect(wrapper.instance().state.hideError).toEqual(false);
    expect(showErrorMessageMethod).toHaveBeenCalledWith(
      "Operation not permitted"
    );
  });

  test("should handle click and value > 0", () => {
    expect.assertions(3);

    props = { ...props, RatesReducer: { exchangesRates: [] } };
    const wrapper = shallow(<App {...props} />);
    wrapper.instance().setState({
      currencyAmounts: {
        GBP: 10,
        USD: 20
      },
      inputValues: { 0: 10.0, 1: -10.0 }
    });

    wrapper.instance().exchangeClickHandler();

    expect(wrapper.instance().state.hideError).toEqual(true);
    expect(wrapper.instance().state.inputValues).toEqual({
      0: "0.00",
      1: "0.00"
    });
    expect(wrapper.instance().state.currencyAmounts).toEqual({
      GBP: 20,
      USD: 10
    });
  });

  test("test mapStateToProps", () => {
    expect.assertions(1);

    const wrapper = mapStateToProps({ RatesReducer: [1234] });

    expect(wrapper).toEqual({ RatesReducer: [1234] });
  });
});
