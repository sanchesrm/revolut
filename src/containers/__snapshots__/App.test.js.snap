// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component render correctly 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#1881ec",
      "height": "100%",
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <Container
    style={
      Object {
        "marginTop": "100px",
        "width": "600px",
      }
    }
    textAlign="center"
  >
    <Header
      style={
        Object {
          "marginBottom": "15px",
        }
      }
      textAlign="center"
    >
      <Image
        as="img"
        src="logo.png"
        style={
          Object {
            "width": "150px",
          }
        }
        ui={true}
      />
    </Header>
    <Connect(CurrencyContent)
      amount={21.23}
      changeHandler={[Function]}
      choosenCurrency={
        Array [
          "GBP",
          "USD",
        ]
      }
      handleCurrencyChoose={[Function]}
      inputValue="0.00"
      position={0}
    />
    <Connect(CurrencyContent)
      amount={11.71}
      changeHandler={[Function]}
      choosenCurrency={
        Array [
          "GBP",
          "USD",
        ]
      }
      handleCurrencyChoose={[Function]}
      inputValue="0.00"
      position={1}
    />
    <Button
      as="button"
      content="Exchange"
      onClick={[Function]}
      primary={true}
      style={
        Object {
          "backgroundImage": "linear-gradient(to right, rgba(0, 128, 0, 0.5), rgba(0, 128, 0, 0.7)",
          "fontSize": "20px",
          "height": "50px",
          "marginTop": "40px",
          "width": "300px",
        }
      }
    />
    <Message
      color="red"
      error={true}
      hidden={true}
      style={
        Object {
          "marginTop": "20px",
        }
      }
    >
      Operation not permitted
    </Message>
  </Container>
</div>
`;
